spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java

```diff
@@ -369,15 +370,53 @@ public void alterNamespace(String[] namespace, NamespaceChange... changes) throw
   }
 
   @Override
+  // Spark assumes that catalogs CASCADE by default. So we have to eagerly
+  // attempt to drop namespaces and tables, but the CASCADE keyword is still
+  // required to actually drop tables and namespaces as Spark will error out
+  // if any of the recursive deletes are non-empty and the user didn't specify
+  // cascades in their query.
   public boolean dropNamespace(String[] namespace) throws NoSuchNamespaceException {
     if (asNamespaceCatalog != null) {
+      Namespace asNamespace = Namespace.of(namespace);
+      boolean exists = namespaceExists(namespace);
+
+      // Spark only throws the catalyst version of `NoSuchNamespaceException` if the namespace
+      // does not exist AND the user did not specify `IF EXISTS` in their query.
+      //
+      // If the namespace does not exist, but listNamespaces didn't throw an exception,
+      // we know the user used IF EXISTS and can return false early.
+      List<Namespace> subNamespaces;
       try {
-        return asNamespaceCatalog.dropNamespace(Namespace.of(namespace));
+        subNamespaces = asNamespaceCatalog.listNamespaces(asNamespace);
       } catch (org.apache.iceberg.exceptions.NoSuchNamespaceException e) {
         throw new NoSuchNamespaceException(namespace);
       }
-    }
 
+      if (!exists && subNamespaces.size() == 0) {
```

Comment:
Is this possible? If the namespace doesn't exist, how would `subNamespaces` be populated? Wouldn't it throw `NoSuchNamespaceException` above? If so, what is the value of `exists`?

Reply:
If a user calls `DROP NAMESPACE IF EXISTS ns` on a namespace that does not exist, it will not throw `NoSuchNamespaceException`.

That's why I check if it exists first. in that case, subNamespaces will be `new Stringp[0][]` with the new set up.